### PR TITLE
FFTW startup runtime option

### PR DESCRIPTION
**Pull request by KunkelAlexander**
This pull request introduces the runtime option  `OPT__FFTW_STARTUP` that determines whether to use `FFTW_ESTIMATE`, `FFTW_MEASURE` or `FFTW_PATIENT` as default for the FFTW plans used in GAMER. 
The corresponding option can be set in `Input__Parameter` via  `OPT__FFTW_STARTUP = 0/1/2` (estimate, measure, patient (only FFTW3).

### CODE REVIEW

**Comment by KunkelAlexander** [comment]
@hyschive 
I also renamed the fftw3 single and double omp options to make the naming consistent with the rest of gamer. 
In addition, I also added three lines of code that read an MHD related option back from the hdf5 file. I think that someone just forgot to read the parameter back in because it is stored in the hdf5 file properly. If you do not want to meddle with this option, I can delete the corresponding lines of code.

**Comment by KunkelAlexander** [comment]
@hyschive 
This pull request is ready for review.

File: src/Output/Output_DumpData_Total_HDF5.cpp [hunk 1]
```diff
@@ -238,6 +238,7 @@ Procedure for outputting new variables:
 //                2463 : 2023/03/20 --> output FB_SEP_FLUOUT
 //                2464 : 2023/04/27 --> output LIBYT_INTERACTIVE
 //                2465 : 2023/04/29 --> output MU_NORM
+//                2466 : 2023/05/08 --> output OPT__FFTW_STARTUP
```
Comment (hyschive) [inline src/Output/Output_DumpData_Total_HDF5.cpp:241]: Search the keyword `FormatVersion` in this file and update it too.

File: src/Output/Output_DumpData_Total_HDF5.cpp [hunk 2]
```diff
@@ -2352,7 +2353,9 @@ void FillIn_InputPara( InputPara_t &InputPara, const int NFieldStored, char Fiel
 #  ifdef MHD
    InputPara.Opt__InitBFieldByFile   = OPT__INIT_BFIELD_BYFILE;
 #  endif
-
+#  ifdef SUPPORT_FFTW
+   InputPara.Opt__FFTW_Startup       = OPT__FFTW_STARTUP;
```
Comment (hyschive) [inline src/Output/Output_DumpData_Total_HDF5.cpp:2357]: You need to call `H5Tinsert()` to insert this new parameter. Check `Opt__InitBFieldByFile` as an example.

**Review by hyschive — CHANGES_REQUESTED** [review]
@KunkelAlexander Thanks for adding this new feature! I've left some inline comments.

File: src/Main/Main.cpp [hunk 3]
```diff
@@ -174,9 +174,12 @@ double               LB_INPUT__PAR_WEIGHT;
 bool                 OPT__RECORD_LOAD_BALANCE;
 #endif
 bool                 OPT__MINIMIZE_MPI_BARRIER;
+#ifdef SUPPORT_FFTW
+int                  OPT__FFTW_STARTUP;
 #if ( SUPPORT_FFTW == FFTW3 )
-bool                 FFTW3_Double_OMP_Enabled, FFTW3_Single_OMP_Enabled;
-#endif
+bool                 FFTW3_DOUBLE_OMP_ENABLED, FFTW3_SINGLE_OMP_ENABLED;
```
[the remark this answers is not in the document]
Comment (KunkelAlexander) [reply src/Main/Main.cpp:180]: That means I should change it back? No problem. :)

File: include/Macro.h [hunk 4]
```diff
@@ -91,6 +91,10 @@
 #define RNG_GNU_EXT  1
 #define RNG_CPP11    2
 
+// FFTW startup options
+#define FFTW_STARTUP_ESTIMATE 0
+#define FFTW_STARTUP_MEASURE  1
+#define FFTW_STARTUP_PATIENT  2
```
[the remark this answers is not in the document]
Comment (KunkelAlexander) [reply include/Macro.h:97]: Done!

File: src/Init/Init_Load_Parameter.cpp [hunk 5]
```diff
@@ -378,6 +379,9 @@ void Init_Load_Parameter()
 #  ifdef MHD
    ReadPara->Add( "OPT__INIT_BFIELD_BYFILE",    &OPT__INIT_BFIELD_BYFILE,         false,           Useless_bool,  Useless_bool   );
 #  endif
+#  ifdef SUPPORT_FFTW
+   ReadPara->Add( "OPT__FFTW_STARTUP",          &OPT__FFTW_STARTUP,              -1,               0,             NoMax_int      );
```
[the remark this answers is not in the document]
Comment (KunkelAlexander) [reply src/Init/Init_Load_Parameter.cpp]: I agree and changed the code accordingly.

File: example/test_problem/Template/Input__Parameter [hunk 6]
```diff
@@ -301,7 +301,7 @@ OPT__INIT_RESTRICT            1           # restrict all data during the initial
 OPT__INIT_GRID_WITH_OMP       1           # enable OpenMP when assigning the initial condition of each grid patch [1]
 OPT__GPUID_SELECT            -1           # GPU ID selection mode: (-3=Laohu, -2=CUDA, -1=MPI rank, >=0=input) [-1]
 INIT_SUBSAMPLING_NCELL        0           # perform sub-sampling during initialization: (0=off, >0=# of sub-sampling cells) [0]
-
+OPT__FFTW_STARTUP            -1           # initialise fftw plans (-1=DEFAULT (MEASURE), 0=ESTIMATE, 1=MEASURE, 2=PATIENT (only FFTW3))
```
[the remark this answers is not in the document]
Comment (KunkelAlexander) [reply example/test_problem/Template/Input__Parameter:304]: Done

File: src/Init/Init_FFTW.cpp [hunk 7]
```diff
@@ -125,6 +125,19 @@ void Init_FFTW()
...
+
+   switch ( OPT__FFTW_STARTUP )
+   {
+      case FFTW_STARTUP_ESTIMATE:    StartupFlag = FFTW_ESTIMATE;               break;
+      case FFTW_STARTUP_MEASURE:     StartupFlag = FFTW_MEASURE;                break;
+#     if ( SUPPORT_FFTW == FFTW3 )
+      case FFTW_STARTUP_PATIENT:     StartupFlag = FFTW_PATIENT;                break;
+#     endif // # if ( SUPPORT_FFTW == FFTW3 )
+
+      default:                       StartupFlag = FFTW_MEASURE;
```
[the remark this answers is not in the document]
Comment (KunkelAlexander) [reply src/Init/Init_FFTW.cpp:139]: Okay. :)

**Review by hyschive — APPROVED** [review]

**Pull request merged**